easybuild/easyblocks/e/elpa.py

```diff
@@ -0,0 +1,200 @@
+##
+# Copyright 2009-2019 Ghent University
+# Copyright 2019 Micael Oliveira
+#
+# This file is part of EasyBuild,
+# originally created by the HPC team of Ghent University (http://ugent.be/hpc/en),
+# with support of Ghent University (http://ugent.be/hpc),
+# the Flemish Supercomputer Centre (VSC) (https://www.vscentrum.be),
+# Flemish Research Foundation (FWO) (http://www.fwo.be/en)
+# and the Department of Economy, Science and Innovation (EWI) (http://www.ewi-vlaanderen.be/en).
+#
+# https://github.com/easybuilders/easybuild
+#
+# EasyBuild is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation v2.
+#
+# EasyBuild is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with EasyBuild.  If not, see <http://www.gnu.org/licenses/>.
+##
+"""
+EasyBuild support for building and installing ELPA, implemented as an easyblock
+
+@author: Micael Oliveira (MPSD-Hamburg)
+"""
+import os
+
+from vsc.utils.missing import nub
+
+from easybuild.easyblocks.generic.configuremake import ConfigureMake
+from easybuild.framework.easyconfig import CUSTOM
+from easybuild.tools.build_log import EasyBuildError
+from easybuild.tools.config import build_option
+from easybuild.tools.systemtools import get_cpu_features, get_shared_lib_ext
+from easybuild.tools.toolchain.compiler import OPTARCH_GENERIC
+ELPA_CPU_FEATURE_FLAGS = ['avx', 'avx2', 'avx512', 'vsx', 'sse4_2']
+
+
+class EB_ELPA(ConfigureMake):
+    """Support for building/installing ELPA."""
+
+    @staticmethod
+    def extra_options():
+        """Custom easyconfig parameters for ELPA."""
+        extra_vars = {
+            'auto_detect_cpu_features': [True, "Auto-detect available CPU features, and configure accordingly", CUSTOM],
+            'with_mpi': [True, "Enable building of ELPA MPI library", CUSTOM],
+            'with_openmp': [True, "Enable building of ELPA OpenMP library", CUSTOM],
```

Comment:
@micaeljtoliveira We have `openmp`, `usempi`, `shared` in `toolchainopts`, better rely on that rather than introducing custom easyconfig parameters specific to ELPA (where possible)? See https://easybuild.readthedocs.io/en/latest/version-specific/toolchain_opts.html#foss

Reply:
@boegel I based my implementation on the FFTW easyblock, which provides these parameters, In any case, I don't see any problem in removing them.

Reply:
The toolchainopt `usempi` may  have an unwanted side-effect (particuarly in the serial case), it switches `CC` to `mpicc` (from `gcc` for `foss`). Given that you are building all the possibilities in different directories, I'd be more inclined to keep these options as a feature of ELPA (they don't really have the same purpose as the `toolchainopts` which are about influencing a particular build, here you are selecting whether or not to *include* a particular build)

Reply:
@ocaisa Thanks for the explanation on the purpose of the `toolchainopts`. That clarifies the logic behind some choices in the easyblocks that were not very clear to me.